evalml/tests/conftest.py

```diff
@@ -1088,3 +1094,217 @@ def _imbalanced_data_X_y(problem_type, categorical_columns, size):
         return X, y
 
     return _imbalanced_data_X_y
+
+
+class _AutoMLTestEnv:
```

Comment:
Please feel free to ignore this comment entirely. But...

You have me wondering how hard it would be to separate the automl-specific parts of this fixture out from the mocking/context parts. And then put up a PR into pytest with the mocking fixture, and several examples of how to use it in different contexts.

Again, not at all necessary for this PR, just food for thought. It could be real cool to give people a tool which makes it easy to define fixtures like this which perform mocking!!!!

Reply:
This is such a cool idea! I think the one thing that we need an abstraction to create multiple nested contexts so that it can be more general. Right now, we're manually combining all the patches into the same context, e.g. `with mock_fit, mock_score, ...`. 

I looked at [ExitStack](https://docs.python.org/3/library/contextlib.html#contextlib.ExitStack) when I was first designing this but it wasn't working like I wanted it to. Could be that I was not using it correctly! So I'll get this merged in and think about this further.